File: random-forest/random_forest.js
class SimpleDecisionTree {
    constructor(maxDepth = 3) {
        this.maxDepth = maxDepth;
    }
    
    gini(y) {
        const counts = {};
        y.forEach(label => counts[label] = (counts[label] || 0) + 1);
        let impurity = 1.0;
        Object.values(counts).forEach(count => {
            const prob = count / y.length;
            impurity -= prob * prob;
        });
        return impurity;
    }
    
    fit(X, y, depth = 0) {
        if (depth >= this.maxDepth || new Set(y).size === 1) {
            return y.reduce((a, b, i, arr) => 
                arr.filter(v => v === a).length >= arr.filter(v => v === b).length ? a : b);
        }
        
        let bestGini = Infinity;
        let bestFeature = null, bestThreshold = null;
        
        // Random feature selection
        const nFeatures = Math.floor(Math.sqrt(X[0].length));
        const features = [];
        while (features.length < nFeatures) {
            const f = Math.floor(Math.random() * X[0].length);
            if (!features.includes(f)) features.push(f);
        }
        
        for (const feature of features) {
            const thresholds = [...new Set(X.map(x => x[feature]))];
            for (const threshold of thresholds) {
                const leftMask = X.map(x => x[feature] <= threshold);
                const yLeft = y.filter((_, i) => leftMask[i]);
                const yRight = y.filter((_, i) => !leftMask[i]);
                
                if (yLeft.length === 0 || yRight.length === 0) continue;
                
                const gini = (yLeft.length * this.gini(yLeft) + yRight.length * this.gini(yRight)) / y.length;
                if (gini < bestGini) {
                    bestGini = gini;
                    bestFeature = feature;
                    bestThreshold = threshold;
                }
            }
        }
        
        if (bestFeature === null) {
            return y.reduce((a, b, i, arr) => 
                arr.filter(v => v === a).length >= arr.filter(v => v === b).length ? a : b);
        }
        
        const leftMask = X.map(x => x[bestFeature] <= bestThreshold);
        const XLeft = X.filter((_, i) => leftMask[i]);
        const XRight = X.filter((_, i) => !leftMask[i]);
        const yLeft = y.filter((_, i) => leftMask[i]);
        const yRight = y.filter((_, i) => !leftMask[i]);
        
        this.feature = bestFeature;
        this.threshold = bestThreshold;
        this.left = new SimpleDecisionTree(this.maxDepth).fit(XLeft, yLeft, depth + 1);
        this.right = new SimpleDecisionTree(this.maxDepth).fit(XRight, yRight, depth + 1);
        return this;
    }
    
    predictOne(x) {
        if (this.feature === undefined) return this;
        
        if (x[this.feature] <= this.threshold) {
            return typeof this.left === 'object' ? this.left.predictOne(x) : this.left;
        } else {
            return typeof this.right === 'object' ? this.right.predictOne(x) : this.right;
        }
    }
}

class RandomForest {
    constructor(nTrees = 10, maxDepth = 3) {
        this.nTrees = nTrees;
        this.maxDepth = maxDepth;
        this.trees = [];
    }
    
    fit(X, y) {
        this.trees = [];
        for (let i = 0; i < this.nTrees; i++) {
            // Bootstrap sampling
            const indices = Array.from({length: X.length}, () => Math.floor(Math.random() * X.length));
            const XSample = indices.map(i => X[i]);
            const ySample = indices.map(i => y[i]);
            
            const tree = new SimpleDecisionTree(this.maxDepth);
            tree.fit(XSample, ySample);
            this.trees.push(tree);
        }
    }
    
    predict(X) {
        return X.map(x => {
            const treePredictions = this.trees.map(tree => tree.predictOne(x));
            const counts = {};
            treePredictions.forEach(pred => counts[pred] = (counts[pred] || 0) + 1);
            return parseInt(Object.keys(counts).reduce((a, b) => counts[a] > counts[b] ? a : b));
        });
    }
}

// Test
const X = [[1, 2], [2, 3], [3, 1], [6, 5], [7, 7], [8, 6]];
const y = [0, 0, 0, 1, 1, 1];

const model = new RandomForest(5);
model.fit(X, y);
console.log(model.predict([[2, 2], [7, 6]])); // [0, 1]

File: random-forest/RandomForest.kt
import kotlin.math.*
import kotlin.random.Random

class SimpleDecisionTree(private val maxDepth: Int = 3) {
    private var feature: Int? = null
    private var threshold: Double? = null
    private var left: Any? = null
    private var right: Any? = null
    
    private fun gini(y: IntArray): Double {
        val counts = y.groupBy { it }.mapValues { it.value.size }
        var impurity = 1.0
        counts.values.forEach { count ->
            val prob = count.toDouble() / y.size
            impurity -= prob * prob
        }
        return impurity
    }
    
    fun fit(X: Array<DoubleArray>, y: IntArray, depth: Int = 0): Any {
        if (depth >= maxDepth || y.distinct().size == 1) {
            return y.groupBy { it }.maxByOrNull { it.value.size }?.key ?: y[0]
        }
        
        var bestGini = Double.POSITIVE_INFINITY
        var bestFeature: Int? = null
        var bestThreshold: Double? = null
        
        // Random feature selection
        val nFeatures = sqrt(X[0].size.toDouble()).toInt()
        val features = (0 until X[0].size).shuffled().take(nFeatures)
        
        for (feature in features) {
            val thresholds = X.map { it[feature] }.distinct()
            for (threshold in thresholds) {
                val leftMask = X.map { it[feature] <= threshold }
                val yLeft = y.filterIndexed { i, _ -> leftMask[i] }.toIntArray()
                val yRight = y.filterIndexed { i, _ -> !leftMask[i] }.toIntArray()
                
                if (yLeft.isEmpty() || yRight.isEmpty()) continue
                
                val gini = (yLeft.size * gini(yLeft) + yRight.size * gini(yRight)) / y.size
                if (gini < bestGini) {
                    bestGini = gini
                    bestFeature = feature
                    bestThreshold = threshold
                }
            }
        }
        
        if (bestFeature == null) {
            return y.groupBy { it }.maxByOrNull { it.value.size }?.key ?: y[0]
        }
        
        val leftMask = X.map { it[bestFeature] <= bestThreshold!! }
        val XLeft = X.filterIndexed { i, _ -> leftMask[i] }.toTypedArray()
        val XRight = X.filterIndexed { i, _ -> !leftMask[i] }.toTypedArray()
        val yLeft = y.filterIndexed { i, _ -> leftMask[i] }.toIntArray()
        val yRight = y.filterIndexed { i, _ -> !leftMask[i] }.toIntArray()
        
        this.feature = bestFeature
        this.threshold = bestThreshold
        this.left = SimpleDecisionTree(maxDepth).fit(XLeft, yLeft, depth + 1)
        this.right = SimpleDecisionTree(maxDepth).fit(XRight, yRight, depth + 1)
        
        return this
    }
    
    fun predictOne(x: DoubleArray): Int {
        if (feature == null) return this as Int
        
        return if (x[feature!!] <= threshold!!) {
            when (left) {
                is SimpleDecisionTree -> left!!.predictOne(x)
                else -> left as Int
            }
        } else {
            when (right) {
                is SimpleDecisionTree -> right!!.predictOne(x)
                else -> right as Int
            }
        }
    }
}

class RandomForest(private val nTrees: Int = 10, private val maxDepth: Int = 3) {
    private val trees = mutableListOf<SimpleDecisionTree>()
    
    fun fit(X: Array<DoubleArray>, y: IntArray) {
        trees.clear()
        
        repeat(nTrees) {
            // Bootstrap sampling
            val indices = IntArray(X.size) { Random.nextInt(X.size) }
            val XSample = indices.map { X[it] }.toTypedArray()
            val ySample = indices.map { y[it] }.toIntArray()
            
            val tree = SimpleDecisionTree(maxDepth)
            val result = tree.fit(XSample, ySample)
            if (result is SimpleDecisionTree) {
                trees.add(result)
            }
        }
    }
    
    fun predict(X: Array<DoubleArray>): IntArray {
        return X.map { x ->
            val treePredictions = trees.map { it.predictOne(x) }
            treePredictions.groupBy { it }.maxByOrNull { it.value.size }?.key ?: 0
        }.toIntArray()
    }
}

fun main() {
    val X = arrayOf(
        doubleArrayOf(1.0, 2.0), doubleArrayOf(2.0, 3.0), doubleArrayOf(3.0, 1.0),
        doubleArrayOf(6.0, 5.0), doubleArrayOf(7.0, 7.0), doubleArrayOf(8.0, 6.0)
    )
    val y = intArrayOf(0, 0, 0, 1, 1, 1)
    
    val model = RandomForest(5, 3)
    model.fit(X, y)
    val predictions = model.predict(arrayOf(doubleArrayOf(2.0, 2.0), doubleArrayOf(7.0, 6.0)))
    println(predictions.contentToString()) // [0, 1]
}
